pkg/controller/meta_controller.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	pb "github.com/risingwavelabs/risingwave-operator/pkg/controller/proto"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+// MetaPodController reconciles a Pod object.
+type MetaPodController struct {
+	client.Client
+	// Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;update;patch
+
+const (
+	metaLeaderLabel = "meta-is-leader"
+)
+
+// metaIsLeader sends a heartbeat to a meta node at ip:port. If meta responds it is the leader.
+func (r *MetaPodController) metaIsLeader(ctx context.Context, ip string, port uint) bool {
+	addr := fmt.Sprintf("%s:%v", ip, port)
+	log := log.FromContext(ctx)
+
+	// TODO: Do we need to make this connection secure?
+	conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		log.Info("did not connect: %v", err)
+		// TODO: retry request
+		return false
+	}
+	defer conn.Close()
+	c := pb.NewHeartbeatServiceClient(conn)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	// TODO: retry this?
+	_, err = c.Heartbeat(ctx, &pb.HeartbeatRequest{})
+	return err != nil
+}
+
+/*
+- Tutorial: https://grpc.io/docs/languages/go/quickstart/
+- Generate proto files via
+
+	protoc --go_out=. --go_opt=paths=source_relative \
+	    --go-grpc_out=. --go-grpc_opt=paths=source_relative \
+	    proto/meta.proto
+*/
+
+// Reconcile handles the pods of the meta service. Will add the metaLeaderLabel to the pods.
+func (r *MetaPodController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// Using a typed object.
+	meta_pods := &corev1.PodList{}
```

Comment:
Be careful about the coding styles.

```suggestion
	metaPods := &corev1.PodList{}
```